File: app/javascript/channels/room_channel.js
import consumer from './consumer'

$(function () {
  const chatChannel = consumer.subscriptions.create({ channel: 'RoomChannel', room: $('#messages').data('room_id') }, {
    connected() {
      // Called when the subscription is ready for use on the server
    },

    disconnected() {
      // Called when the subscription has been terminated by the server
    },

    received: function (data) {
      return $('#messages').append(data['message']);
    },

    speak: function (message) {
      return this.perform('speak', {
        message: message
      });
    }
  });

  $(document).on('keypress', '[data-behavior~=room_speaker]', function (event) {
    if (event.keyCode === 13) {
      const message = event.target.value;
      const roomId = $('input:hidden[name="room"]').val()

      chatChannel.speak({ message: message, room_id: roomId});
      event.target.value = '';
      return event.preventDefault();
    }
  });
});

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from 'room_channel'
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(payload)
    Message.create!(user_id: User.first.id, room_id: payload['message']['room_id'], content: payload['message']['message'])
  end
end
